crates/sui-core/src/state_accumulator.rs

```diff
@@ -98,45 +115,88 @@ impl StateAccumulator {
             .flat_map(|fx| {
-                fx.modified_at_versions()
+                fx.unwrapped()
+                    .iter()
+                    .map(|(oref, _owner)| (oref.0, oref.1))
+                    .collect::<Vec<(ObjectID, SequenceNumber)>>()
+            })
+            .chain(effects.iter().flat_map(|fx| {
+                fx.unwrapped_then_deleted()
                     .iter()
-                    .map(|(id, seq_num)| ObjectKey(*id, *seq_num))
-                    .collect::<Vec<ObjectKey>>()
+                    .map(|oref| (oref.0, oref.1))
+                    .collect::<Vec<(ObjectID, SequenceNumber)>>()
+            }))
+            .collect::<Vec<(ObjectID, SequenceNumber)>>();
+
+        let unwrapped_ids: HashSet<ObjectID> =
+            all_unwrapped.iter().map(|(id, _)| id).cloned().collect();
+
+        // Collect keys from modified_at_versions to remove from the accumulator.
+        // Filter all unwrapped objects (from unwrapped or unwrapped_then_deleted effects)
+        // as these were inserted into the accumulator as a WrappedObject. Will handle these
+        // separately.
+        let modified_at_version_keys: Vec<ObjectKey> = effects
+            .iter()
+            .flat_map(|fx| fx.modified_at_versions())
+            .filter_map(|(id, seq_num)| {
+                if unwrapped_ids.contains(id) {
+                    None
+                } else {
+                    Some(ObjectKey(*id, *seq_num))
+                }
             })
             .collect();
 
         let modified_at_digests: Vec<_> = self
             .authority_store
-            .multi_get_object_by_key(&modified_at_version_keys)
+            .multi_get_object_by_key(&modified_at_version_keys.clone())
             .expect("Failed to get modified_at_versions object from object table")
             .into_iter()
-            .map(|obj| {
-                obj.expect(
-                    "Object from modified_at_versions effects does not exist in objects table",
-                )
+            .zip(modified_at_version_keys)
+            .map(|(obj, key)| {
+                obj.unwrap_or_else(|| panic!("Object for key {:?} from modified_at_versions effects does not exist in objects table", key))
                 .compute_object_reference()
                 .2
             })
             .collect();
-
         acc.remove_all(modified_at_digests);
 
-        epoch_store.insert_state_hash_for_checkpoint(&checkpoint_seq_num, &acc)?;
-        debug!("Accumulated checkpoint {}", checkpoint_seq_num);
+        // Process unwrapped and unwrapped_then_deleted effects, which need to be
+        // removed as WrappedObject using the last sequence number it was tombstoned
+        // against. Unfortunately since this happened in a past transaction, and the
+        // child object may have been modified since (and hence its sequence number incremented),
+        // we must traverse the objects table until we find the tombstone in order to get the
```

Comment:
this comment is a bit out of date now - we don't really traverse, we can seek right to it